provider/boltmem/boltmem.go

```diff
@@ -0,0 +1,426 @@
+		_, err := tx.CreateBucketIfNotExists(bktAlerts)
+		return err
+	})
+	return &Alerts{
+		db:        db,
+		listeners: map[int]chan *types.Alert{},
+		next:      0,
+	}, err
+}
+
+func (a *Alerts) Close() error {
+	return a.db.Close()
+}
+
+// Subscribe returns an iterator over active alerts that have not been
+// resolved and successfully notified about.
+// They are not guaranteed to be in chronological order.
+func (a *Alerts) Subscribe() provider.AlertIterator {
+	var (
+		ch   = make(chan *types.Alert, 200)
+		done = make(chan struct{})
+	)
+	alerts, err := a.getPending()
+
+	i := a.next
+	a.next++
+
+	a.mtx.Lock()
+	a.listeners[i] = ch
+	a.mtx.Unlock()
+
+	go func() {
+		defer func() {
+			a.mtx.Lock()
+			delete(a.listeners, i)
+			close(ch)
+			a.mtx.Unlock()
+		}()
+
+		for _, a := range alerts {
+			select {
+			case ch <- a:
+			case <-done:
+				return
+			}
+		}
+
+		<-done
+	}()
+
+	return provider.NewAlertIterator(ch, done, err)
+}
+
+// GetPending returns an iterator over all alerts that have
+// pending notifications.
+func (a *Alerts) GetPending() provider.AlertIterator {
+	var (
+		ch   = make(chan *types.Alert, 200)
+		done = make(chan struct{})
+	)
+
+	alerts, err := a.getPending()
+
+	go func() {
+		defer close(ch)
+
+		for _, a := range alerts {
+			select {
+			case ch <- a:
+			case <-done:
+				return
+			}
+		}
+	}()
+
+	return provider.NewAlertIterator(ch, done, err)
+}
+
+func (a *Alerts) getPending() ([]*types.Alert, error) {
+	var alerts []*types.Alert
+
+	err := a.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket(bktAlerts)
+		c := b.Cursor()
+
+		for k, v := c.First(); k != nil; k, v = c.Next() {
+			var a types.Alert
+			if err := json.Unmarshal(v, &a); err != nil {
+				return err
+			}
+			alerts = append(alerts, &a)
+		}
+
+		return nil
+	})
+	return alerts, err
+}
+
+// Get returns the alert for a given fingerprint.
+func (a *Alerts) Get(fp model.Fingerprint) (*types.Alert, error) {
+	var alert types.Alert
+	err := a.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket(bktAlerts)
+
+		fpb := make([]byte, 8)
+		binary.BigEndian.PutUint64(fpb, uint64(fp))
+
+		ab := b.Get(fpb)
+		if ab == nil {
+			return provider.ErrNotFound
+		}
+
+		return json.Unmarshal(ab, &alert)
+	})
+	return &alert, err
+}
+
+// Put adds the given alert to the set.
+func (a *Alerts) Put(alerts ...*types.Alert) error {
+	a.mtx.Lock()
+	defer a.mtx.Unlock()
+
+	err := a.db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket(bktAlerts)
+
+		for _, alert := range alerts {
+			fp := make([]byte, 8)
+			binary.BigEndian.PutUint64(fp, uint64(alert.Fingerprint()))
+
+			ab := b.Get(fp)
+
+			// Merge the alert with the existing one.
+			if ab != nil {
+				var old types.Alert
+				if err := json.Unmarshal(ab, &old); err != nil {
+					return fmt.Errorf("decoding alert failed: %s", err)
+				}
+				// Merge alerts if there is an overlap in activity range.
+				if (alert.EndsAt.Sub(old.StartsAt) > 0 && alert.EndsAt.Sub(old.EndsAt) < 0) ||
```

Comment:
As we are talking about nanoseconds in this case it doesn't really matter.
But I can add that later. Not a blocker for testing.